### PR TITLE
feat(WMS): call getCapabilities during layer preprocessing

### PR DESCRIPTION
**Pull request by autra**
## Description

Add GetCapabilities query for WMS layers. This PR implement the fetching of capabilities and set or check the following informations:
- projection
- extent
- format

BREAKING CHANGE:

- WMS layer will now query getCapabilities. If you want to disable this  behaviour (because you already provide all the informations in the json object), you need to set layer.disableGetCap to true.

- previously we had layer.format and layer.options.mimetype coexisting, with no clear difference in role. This commit removes layer.options.mimetype to only keep layer.format. This is needed to reliably check configured format.

## Motivation and Context

Supporting getCapabilities in our providers is a necessary step to present a usable SIG application for the outside world, let's start this.

Moreover, if we want itowns to be able to display a layer from its url only, we need to be able to query capabilities from these layers.

Work that would need to follow this PR:
- support getCapabilities for WFS and WMTS 
- add the possibility to create an entire scene (view + layers) with just a WMS/WMTS/WFS url.

This is a first pass at fixing #285.

### CODE REVIEW

**Comment by autra** [comment]
@iTowns/reviewers please r?

**Review by gchoqueux — CHANGES_REQUESTED** [review]
We need to valid the default value of `layer.disableGetCap`.

Why you change  `options.mimetype` to format?
You have to support both if you make this type of change. (and tag to deprecated)

**Comment by autra** [comment]
> Why you change options.mimetype to format?

From the description

> previously we had layer.format and layer.options.mimetype coexisting, with no clear difference in role. This commit removes layer.options.mimetype to only keep layer.format. This is needed to reliably check configured format.

2 option for the same thing is not needed. I picked the better.

>  You have to support both if you make this type of change. (and tag to deprecated)

For such a simple thing? Deprecation process is necessary if the change will notably impact client, preventing them to work before extensive work is done to adapt themselves to a new version. It is not necessary to go into such a process for a simple renaming. As the commit is annotated with a `BREAKING CHANGE`, it'll appear automatically in the release note. I'll make sure it is clearly visible.

> We need to valid the default value of layer.disableGetCap.

It is very surprising that itowns managed to do *without* getCap this far honestly :-) Every SIG app I'm aware of do that by default. Of course we'll keep a way to bypass it (hence the disableGetCap, the name can be changed btw). It is very useful in some case I agree (for instance for Lyon data, their getCap takes 17 sec :-/ )

I believe this will make iTowns a lot simpler to use, especially for WMTS layers.

File: src/Core/Geographic/Coordinates.js [hunk 1]
```diff
@@ -10,6 +10,8 @@ import mE from '../Math/MathExtended';
 import Ellipsoid from '../Math/Ellipsoid';
 
 proj4.defs('EPSG:4978', '+proj=geocent +datum=WGS84 +units=m +no_defs');
+// used in some protocol
+proj4.defs('CRS:84', proj4.defs('EPSG:4326'));
```
Comment (gchoqueux) [inline src/Core/Geographic/Coordinates.js:14]: can you add comment?
```js
// CRS:84 is equivalent to EPSG:4326 - ie, basic WGS84 degrees.
```

File: src/Core/Scheduler/Providers/WMS_Provider.js [hunk 2]
```diff
@@ -7,6 +7,78 @@
...
+/* For reference, there are also these 2 possible tags that do not translate to mimetype
+   INIMAGE display text in the returned image -->
+   BLANK return an image with all pixels transparent if
+   */
+
+function _parseSupportedCrs(version, xmlLayer) {
+    let supportedCrs = [];
+    const crsPropName = version === '1.3.0' ? 'CRS' : 'SRS';
+
+    for (const childElem of xmlLayer.children) {
```
Comment (gchoqueux) [inline src/Core/Scheduler/Providers/WMS_Provider.js]: Searching for items by tag name is often used. It could be factoring.

Otherwise, it will not be easier to transform the xml object js?
I also found lib js that turns the capes into json.

**Comment by autra** [comment]
> We need to valid the default value of layer.disableGetCap.

Ok let's validate it. What's your thoughts about it?

File: src/Core/Scheduler/Providers/WMS_Provider.js [hunk 3]
```diff
@@ -7,6 +7,78 @@
...
+                    childElem.getAttribute('maxy'));
+        }
+    }
+    if (xmlLayer.parentNode && xmlLayer.parentNode.tagName === 'Layer') {
+        return parseLayerBoundingBox(version, targetCrs, xmlLayer.parentNode);
+    }
+}
+
+function parseGeoBounds(version, xmlLayer) {
+    const geoTagName = version === '1.3.0' ? 'EX_GeographicBoundingBox' : 'LatLonBoundingBox';
```
Comment (gchoqueux) [inline src/Core/Scheduler/Providers/WMS_Provider.js]: you could group in private functions 
```js
getCrsPNByVersion(version)...

getGeoPNByVersion(version)...
```

File: src/Core/Scheduler/Providers/WMS_Provider.js [hunk 4]
```diff
@@ -45,53 +117,155 @@ WMS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer)
...
+        }&FORMAT=${layer.format
+        }&TRANSPARENT=${layer.transparent
+        }&BBOX=%bbox` +
+        `&${crsPropName}=${layer.projection
+        }&WIDTH=${layer.width
+        }&HEIGHT=${layer.width}`;
+    });
+};
+
+WMS_Provider.prototype.findXmlLayer = function findXMLLayer(name, xml) {
```
Comment (gchoqueux) [inline src/Core/Scheduler/Providers/WMS_Provider.js]: Could you add documentation on new functions?

File: src/Core/Scheduler/Providers/WMS_Provider.js [hunk 5]
```diff
@@ -45,53 +117,155 @@ WMS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer)
...
+            return n.parentNode;
+        }
+    }
+};
+
+WMS_Provider.prototype.parseSupportedCrs = function parseSupportedCrs(version, xmlLayer) {
+    return new Set(_parseSupportedCrs(version, xmlLayer));
+};
+
+WMS_Provider.prototype.parseSupportedFormats = function parseSupportedFormats(version, xmlCapa) {
```
Comment (gchoqueux) [inline src/Core/Scheduler/Providers/WMS_Provider.js]: There's a difference between `xmlLayer` and `xmlCapa`

File: src/Core/Scheduler/Providers/WMTS_Provider.js [hunk 6]
```diff
@@ -21,18 +21,18 @@ WMTS_Provider.prototype.customUrl = function customUrl(layer, url, tilematrix, r
...
         const options = layer.options;
         options.version = options.version || '1.0.0';
         options.tileMatrixSet = options.tileMatrixSet || 'WGS84';
-        options.mimetype = options.mimetype || 'image/png';
         options.style = options.style || 'normal';
         options.projection = options.projection || 'EPSG:3857';
         let newBaseUrl = `${layer.url}` +
             `?LAYER=${options.name}` +
-            `&FORMAT=${options.mimetype}` +
+            `&FORMAT=${layer.format}` +
```
Comment (gchoqueux) [inline src/Core/Scheduler/Providers/WMTS_Provider.js]: `options.mimetype` to `layer.format` to be in another PR, both must be supported with a deprecated message

File: src/Core/View.js [hunk 7]
```diff
@@ -296,6 +296,13 @@ View.prototype.addLayer = function addLayer(layer, parentLayer) {
         throw new Error(`${layer.protocol} is not a recognized protocol name.`);
     }
     layer = _preprocessLayer(this, layer, provider);
+    layer.whenReady.then(() => {
+        // if we still haven't any validExtent after getCapabilities, we assume
+        // it is the same as the configured extent
+        if (!layer.validExtent) {
+            layer.validExtent = layer.extent;
```
Comment (gchoqueux) [inline src/Core/View.js]: and if `layer.extent` is undefined?

**Review by gchoqueux — CHANGES_REQUESTED** [review]

File: src/Core/Geographic/Coordinates.js (hunk 1 above)
Comment (autra) [reply src/Core/Geographic/Coordinates.js:14]: I don't see the point of this comment. This line is pretty self-explanatory.

File: src/Core/Scheduler/Providers/WMS_Provider.js (hunk 3 above)
Comment (autra) [reply src/Core/Scheduler/Providers/WMS_Provider.js]: You want a function for a one-liner?

File: src/Core/Scheduler/Providers/WMS_Provider.js (hunk 4 above)
Comment (autra) [reply src/Core/Scheduler/Providers/WMS_Provider.js]: Oups I forgot! yes will do.

File: src/Core/Scheduler/Providers/WMS_Provider.js (hunk 5 above)
Comment (autra) [reply src/Core/Scheduler/Providers/WMS_Provider.js]: Yes

File: src/Core/Geographic/Coordinates.js (hunk 1 above)
Comment (gchoqueux) [reply src/Core/Geographic/Coordinates.js:14]: not for the WGS84

File: src/Core/Scheduler/Providers/WMS_Provider.js (hunk 3 above)
Comment (gchoqueux) [reply src/Core/Scheduler/Providers/WMS_Provider.js]: it's just to group the constants
you can suggest something else

File: src/Core/Scheduler/Providers/WMTS_Provider.js (hunk 6 above)
Comment (autra) [reply src/Core/Scheduler/Providers/WMTS_Provider.js]: I can put that in another PR if you want, but I still don't see the need for a depreciation message. I did give my reasons, but you didn't give yours.

File: src/Core/View.js (hunk 7 above)
Comment (autra) [reply src/Core/View.js]: Normally it shouldn't happen: if disableGetCap is true, we check for it. Otherwise the `parseExtent` method has several fallbacks that cannot reasonably fail.

File: src/Core/Geographic/Coordinates.js (hunk 1 above)
Comment (autra) [reply src/Core/Geographic/Coordinates.js:14]: I see what you mean, but I'm not sure this is the place to explain CRS.

File: src/Core/Scheduler/Providers/WMTS_Provider.js (hunk 6 above)
Comment (gchoqueux) [reply src/Core/Scheduler/Providers/WMTS_Provider.js]: I think it breaks the users cases

File: src/Core/Scheduler/Providers/WMS_Provider.js (hunk 2 above)
Comment (autra) [reply src/Core/Scheduler/Providers/WMS_Provider.js]: > Searching for items by tag name is often used. It could be factoring.

I've hesitated to do so, but each time it's a bit different. So I decided against it because:
- It will only abstract the for loop in the general case (the if part is always a bit different)
- it is less readable than the current code. And it's better to optimize for the reader of our code.

Comment (autra) [reply src/Core/Scheduler/Providers/WMS_Provider.js]: > Otherwise, it will not be easier to transform the xml object js?

In my opinion, not worthy the extra lib. It won't save me the for loop anyway, and it adds a layer of abstraction that doesn't bring a lot of value here.

File: src/Core/Geographic/Coordinates.js (hunk 1 above)
Comment (gchoqueux) [reply src/Core/Geographic/Coordinates.js:14]: I had to watch on the internet for explanations CRS:84, so it's a good place to explain to the developers. Why it makes sense to add this equivalence because we try to support wgs84 in itowns. It's better to be generous in commentary and it'sn't much to add

File: src/Core/Scheduler/Providers/WMTS_Provider.js (hunk 6 above)
Comment (autra) [reply src/Core/Scheduler/Providers/WMTS_Provider.js]: Yes, but what about my suggestion in my previous comment? If it's clearly documented in the release notes, as it is a trivial change to do, user won't have any difficulty right? They would upgrade their itowns lib, and in the same commit rename this property. 

Anyway, I'll see if I can remove this from this PR.

File: src/Core/Geographic/Coordinates.js (hunk 1 above)
Comment (autra) [reply src/Core/Geographic/Coordinates.js:14]: ok

File: src/Core/Scheduler/Providers/WMS_Provider.js (hunk 2 above)
Comment (autra) [reply src/Core/Scheduler/Providers/WMS_Provider.js]: To elaborate, as we already have everything we need to query and parse xml in js, I think we can do without. It will avoid growing itowns.js.

Comment (gchoqueux) [reply src/Core/Scheduler/Providers/WMS_Provider.js]: ok

File: src/Core/Scheduler/Providers/WMS_Provider.js (hunk 3 above)
Comment (autra) [reply src/Core/Scheduler/Providers/WMS_Provider.js]: Oh I see. Yes, a function should be better than constants here I guess.

File: src/Core/Scheduler/Providers/WMTS_Provider.js (hunk 6 above)
Comment (gchoqueux) [reply src/Core/Scheduler/Providers/WMTS_Provider.js]: As it breaks the user cases, it's better, we have to support both and put a deprecated message

Comment (autra) [reply src/Core/Scheduler/Providers/WMTS_Provider.js]: Here it is: https://github.com/iTowns/itowns/pull/597

File: src/Core/Scheduler/Providers/WMS_Provider.js (hunk 4 above)
Comment (autra) [reply src/Core/Scheduler/Providers/WMS_Provider.js]: Actually I did better: I only expose them through a `_testing` export, which avoid polluting the prototype of the provider with private functions.

**Comment by autra** [comment]
@gchoqueux updated.

This now depends on #597

**Comment by autra** [comment]
@iTowns/reviewers updated and rebased :-)

File: examples/wfs.js [hunk 8]
```diff
@@ -31,6 +31,7 @@ view.addLayer({
     networkOptions: { crossOrigin: 'anonymous' },
     type: 'color',
     protocol: 'wms',
+    disableGetCap: true, // their getCap call is too slow
```
Comment (peppsac) [inline examples/wfs.js]: Can't this be implicit? If the user supplies all needed paramaters (format, version?) then you assume that the `GetCapabilities` query is not needed.
WDYT?

Comment (autra) [reply examples/wfs.js]: It's an interesting idea. The advantage of `disableGetCap` is that it *explicitly* ensure getCap won't be called. If it's implicit, it can still be called if you forget something. Maybe we can cover both use cases ? if we have all the parameters, we disable it. If we have disableGetCap, we check the needed parameters and throw if one is missing. WDYT?

Also, WDYT about `disableGetCap` naming? The more I see it, the more I prefer `disableGetCapabilities`.

Comment (peppsac) [reply examples/wfs.js]: Your suggestion seems fine.

And I prefer `disableGetCapabilities` (or `noGetCapabilities`?)

**Comment by autra** [comment]
@peppsac updated
